*: switch to vendored dependencies

dt opened the pull request:
At a high level, this change adds a git submodule at vendor pointing to a new, standalone repo that contains a snapshot of all of our dependencies, then it calls git submodule update --init in the .bootstrap hook that currently maintains our dependencies.

The standalone repo https://github.com/cockroachdb/vendored is currently generated with gvt, but could be generated with anything.

Glock is still being run for now, to install the (vendored) tools we use into GOPATH/bin (though a followup could remove this, installing them to a local bin directory where check could find them, rather than relying on/clobbering the user's PATH).

Updating dependencies would be a two step process: first run gvt (or whatever) to update the dependencies repo and push that ref, then send a pull request to the main repo bumping the submodule to point to the new ref.

The .bootstrap hook doesn't currently run submodule sync, so that those that want to can update the remote url (e.g. to be able to push).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9900)

<!-- Reviewable:end -->

dt committed:
*: add vendor submodule

At a high level, this change adds a git submodule at vendor pointing to a new, standalone repo that contains a snapshot of all of our dependencies, then it calls git submodule update --init in the .bootstrap hook that currently maintains our dependencies.

The standalone repo https://github.com/cockroachdb/vendored is currently generated with glide, but could be generated with anything.

Glock is still being run for now, to install the (vendored) tools we use into GOPATH/bin (though a followup could remove this, installing them to a local bin directory where check could find them, rather than relying on/clobbering the user's PATH).

Updating dependencies would be a two step process: first run glide (or whatever) to update the dependencies repo and push that ref, then send a pull request to the main repo bumping the submodule to point to the new ref.

The .bootstrap hook doesn't currently run submodule sync, so that those that want to can update the remote url (e.g. to be able to push).

Fixes #4182

dt commented:
Blocked on #9844

dt commented:
okay, with #9844 merged, I think this is RFAL

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  :sunglasses: 

---

Review status: 0 of 13 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[build/glock-to-gvt.sh, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDHn_zs6XHgrTe3mk:-KTuDHn_zs6XHgrTe3ml:bn4mzi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/glock-to-gvt.sh#L1)):*
> ```Shell
> #!/bin/bash
> ```

Would be good to add a comment explaining what the script does and for what it should be used (or if it's here just for historic purposes).

---

*[build/setup-gopath.sh, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDZ-KQtTU_45siq8M:-KTuDZ-KQtTU_45siq8N:b-pn9vk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/setup-gopath.sh#L1)):*
> ```Shell
> gvt list | while read -a i; do echo ${i[0]} ${i[3]}; done | glock sync -n
> ```

Add comment explaining what this does and when/why it should be used.

---

*[build/vendor-repo.md, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDhVAU8nk7lP-UgV1:-KTuDhVAU8nk7lP-UgV2:b6vm104) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/vendor-repo.md#L1)):*
> ```Markdown
> # EXPERIMENTAL - DO NOT USE
> ```

Should we remove this "experimental" part?

---

*[build/vendor-repo.md, line 18 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDvqbSgAueKAAoKud:-KTuDvqbSgAueKAAoKue:b647pgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/vendor-repo.md#L18)):*
> ```Markdown
> * update dependencies to their latest version with `gvt update -all`
> 
> ## Working with Submodules
> ```

Maybe this stuff belongs in CONTRIBUTING.md

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KTuEHTx-Cg8EXbBRUKZ:b-j4196b)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 13 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[build/vendor-repo.md, line 16 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuFCrD8QIUZ8v0F_j9:-KTuFCrD8QIUZ8v0F_jA:bqs066o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/vendor-repo.md#L16)):*
> ```Markdown
> * run `gvt` in the parent of this directory (i.e. your `cockroachdb/cockroach` checkout)
> * add new dependencies with `gvt fetch github.com/my/dependency`
> * update dependencies to their latest version with `gvt update -all`
> ```

Is there any advice on how to get all the transitive dependencies? Would we still use glock and then glock-to-gvt.sh?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 14 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[build/glock-to-gvt.sh, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDHn_zs6XHgrTe3mk:-KTuFBp-1UlEfy7VKYKT:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/glock-to-gvt.sh#L1)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Would be good to add a comment explaining what the script does and for what it should be used (or if it's here just for historic purposes).
</details>
Done

---

*[build/setup-gopath.sh, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDZ-KQtTU_45siq8M:-KTuG8gMovhaEhcUrStx:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/setup-gopath.sh#L1)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Add comment explaining what this does and when/why it should be used.
</details>
Done

---

*[build/vendor-repo.md, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDhVAU8nk7lP-UgV1:-KTuFiZ1MF3dD3LPfppt:b6uf8ru) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/vendor-repo.md#L1)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Should we remove this "experimental" part?
</details>
The transition script copies this into `vendor/README` so it serves as the docs on https://github.com/cockroachdb/vendored, so I'm inclined to keep the warning until enable branch protection for `vendored` repo -- I was thinking I'd wait until we're confident this is merging to do that, but could do it now.

---

*[build/vendor-repo.md, line 16 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuFCrD8QIUZ8v0F_j9:-KTuFXURM83lwPNBgrPo:b-cchej9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/vendor-repo.md#L16)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Is there any advice on how to get all the transitive dependencies? Would we still use glock and then glock-to-gvt.sh?
</details>
`gvt fetch` grabs/freezes transitive deps too.

---

*[build/vendor-repo.md, line 18 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDvqbSgAueKAAoKud:-KTuIJdGOUMkLsJiV6lR:b-ckbbsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/vendor-repo.md#L18)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Maybe this stuff belongs in CONTRIBUTING.md
</details>
Added a pointer in  CONTRIBUTING.md  to see `vendor/README.md` for detailed instructions on changing deps.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I'll test it out locally.

---

Reviewed 9 of 13 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[CONTRIBUTING.md, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuYv45VUsaXr2er2fy:-KTuYv45VUsaXr2er2fz:b-jv007f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/CONTRIBUTING.md#L45)):*
> ```Markdown
> - `go generate ./pkg/...` to update generated files -- prefer
>   `go generate ./the-updated-package` instead of `...` when possible to avoid
>   re-generating files in directories where you haven't made any changes.
> ```

Drive-by, feel free to ignore: is this really good advice? It's hard to know what changes, esp. when editing protos or anything that goes to embedded.go. I'd not mention that and rely on advanced users to know what they're doing.

---

*[build/setup-gopath.sh, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDZ-KQtTU_45siq8M:-KTuZmFaetR-ZZrKiAow:bfu9j1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/setup-gopath.sh#L1)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> Done
</details>
`set -euo pipefail`

---

*[build/vendor-repo.md, line 21 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTu_2eW6Q8PDSTwGl6I:-KTu_2eW6Q8PDSTwGl6J:b-uw3ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/vendor-repo.md#L21)):*
> ```Markdown
> Since dependendies are stored in their own repository, and `cockroachdb/cockroach` depends on it,
> changing a dependency requires pushing two things, in order: first the changes to the `vendored`
> respository, then the reference to those changes to the main repository.
> ```

s/resp/rep/

---

*[build/vendor-repo.md, line 26 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTu_V1iOldRa3YX6hNm:-KTu_V1jqJx9LpLrgLjD:bqhra6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/vendor-repo.md#L26)):*
> ```Markdown
> and commit changes (or use `git -C`). The commit must be available on `github.com/cockroachdb/vendored`
> *before* submitting a pull request to `cockroachdb/cockroach` that references it.
> * Organization members can new refs push directly.
> ```

We should have a lint in CI that ensures that the commit which is tagged is actually in the `master` branch. Otherwise, there's a risk of regression as two people bump different dependencies on their branch and accidentally merge that in our main repo (the earlier dependency bumped would be unbumped).

---

*[build/vendor-repo.md, line 38 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTu_Om_qUMjkoCtj0gM:-KTu_Om_qUMjkoCtj0gN:bidel3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/vendor-repo.md#L38)):*
> ```Markdown
> submodule at `./vendor`.
> 
> If a go build fails to find a dependency in `./vendor`, it will continue searching anything named
> ```

Ha, good point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KTu_mjwNoTD8qS926BZ:b-8tt7gg)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[build/vendor-repo.md, line 26 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTu_V1iOldRa3YX6hNm:-KTufUZKndT2IhHUzcr3:bmvk6i7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/vendor-repo.md#L26)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> We should have a lint in CI that ensures that the commit which is tagged is actually in the `master` branch. Otherwise, there's a risk of regression as two people bump different dependencies on their branch and accidentally merge that in our main repo (the earlier dependency bumped would be unbumped).
</details>
I wonder if we should symlink the `manifest` file to a file in the main repo where we linearize history, and then have `gvt restore` (which makes sure everything in manifest is on disk) be part of the workflow for pushing changes to the `vendored` repo.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tamird commented:
I have one high-level question: we're now using TC which has a pseudo-persistent GOPATH. Suppose someone adds a dependency but forgets to go through the motions of adding it to the vendor repo and the mistake makes it through code review - will anything catch it? I think we have this problem today as well, FWIW.

---

Reviewed 9 of 13 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[build/vendor-repo.md, line 26 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTur2YkQw6jVRPvRdoU:-KTur2YkQw6jVRPvRdoV:b-pxdcyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/vendor-repo.md#L26)):*
> ```Markdown
> and commit changes (or use `git -C`). The commit must be available on `github.com/cockroachdb/vendored`
> *before* submitting a pull request to `cockroachdb/cockroach` that references it.
> * Organization members can new refs push directly.
> ```

this sentence is broken

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KTuqIA5dw5caPsgxr2l:b1gzmg8)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[build/glock-to-gvt.sh, line 14 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTw4gTM6Q03yyFVRz4A:-KTw4gTM6Q03yyFVRz4B:b-nk6abo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/glock-to-gvt.sh#L14)):*
> ```Shell
>   cmd)
>     ;;
>   github.com/cockroachdb/c-protobuf)
> ```

Why this special case? It looks just like "*" below.

---

*[build/setup-gopath.sh, line 5 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTw-bkirJIjYfIYSJ-o:-KTw-bkjboTokXwxv04J:b-715bji) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/setup-gopath.sh#L5)):*
> ```Shell
> # This script uses the vendor manifest and glock to setup GOPATH such that it
> # contains all the dependencies at the same revisions as the vendor directory,
> # for use if/when one wants a managed GOPATH.
> ```

You can't switch completely to a managed GOPATH with this script because the vendor paths are hard-coded (in cgo flags at least). I think in practice you'll only be able to use this for selected deps instead of everything, so I'm not sure if this script is that helpful. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KTw4sthVDxiU4tJxvM8:bb6ywmx)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 10 of 14 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[CONTRIBUTING.md, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuYv45VUsaXr2er2fy:-KTwLqBRwcqfn5VoKyi7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/CONTRIBUTING.md#L45)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Drive-by, feel free to ignore: is this really good advice? It's hard to know what changes, esp. when editing protos or anything that goes to embedded.go. I'd not mention that and rely on advanced users to know what they're doing.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 6 of 15 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[build/vendor-repo.md, line 26 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTu_V1iOldRa3YX6hNm:-KU2aW2AaoW5jQqjFWgR:b3liuek) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/vendor-repo.md#L26)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> I wonder if we should symlink the `manifest` file to a file in the main repo where we linearize history, and then have `gvt restore` (which makes sure everything in manifest is on disk) be part of the workflow for pushing changes to the `vendored` repo.
</details>
Yeah, I ended up deciding this to go this route -- `manifest` is now a symlink back up into the main repo, so that should force it to linearize edit history in the main repo. After this merges we can add a tc build that confirms a`gvt restore` (which refetches everything in the manifest) changes nothing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 6 of 15 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[build/vendor-repo.md, line 26 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTu_V1iOldRa3YX6hNm:-KU2j97tj2OMHucgTDFq:bw2qq9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/vendor-repo.md#L26)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> Yeah, I ended up deciding this to go this route -- `manifest` is now a symlink back up into the main repo, so that should force it to linearize edit history in the main repo. After this merges we can add a tc build that confirms a`gvt restore` (which refetches everything in the manifest) changes nothing.
</details>
What's the benefit to simply checking in `cockroachdb/cockroach`'s CI that the committed submodule is in `vendored`'s `master`? I'm not keen on having CI refetch the world in each run.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[build/vendor-repo.md, line 26 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTu_V1iOldRa3YX6hNm:-KU2l4ZfZMRDNI4r0afk:b-kzigok) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/vendor-repo.md#L26)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What's the benefit to simply checking in `cockroachdb/cockroach`'s CI that the committed submodule is in `vendored`'s `master`? I'm not keen on having CI refetch the world in each run.
</details>


From what I understand, the motivation for requiring reachability from master is a) to make sure we don't clobber each other's changes and b) to ensure any referenced dep snapshot won't be GC'ed. 

As far as a), I'm not entirely convinced that just reachability from master is sufficient:
since the dep tree is mechanically generated, people are going to be accustomed to just blindly committing all edits. While it seems unlikely someone would do anything other than just re-run `gvt restore` to resolve a merge conflict, it's very cheap to guard against -- refetching, from a machine in the cloud, should be pretty cheap, and can be a parallel ci task, so it does't need to add any extra time to a build finishing. The fact that edits are pushed here _before_ they're linearized in the main repo history (since they PR needs to reference a SHA) makes me even more worried that it could drift from the canonical manifest. 

And as far as b), we get that with reachability from master + gh branch protection, but I might want to go ahead and tag `vendored` for every released cockroach SHA as well, just to be safe.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[build/vendor-repo.md, line 26 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTu_V1iOldRa3YX6hNm:-KU2te1jJTx1Qie09xVa:b-1lshby) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4bb0b6e4522c873c1920bdef2184d407837409/build/vendor-repo.md#L26)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> 
> 
> From what I understand, the motivation for requiring reachability from master is a) to make sure we don't clobber each other's changes and b) to ensure any referenced dep snapshot won't be GC'ed. 
> 
> As far as a), I'm not entirely convinced that just reachability from master is sufficient:
> since the dep tree is mechanically generated, people are going to be accustomed to just blindly committing all edits. While it seems unlikely someone would do anything other than just re-run `gvt restore` to resolve a merge conflict, it's very cheap to guard against -- refetching, from a machine in the cloud, should be pretty cheap, and can be a parallel ci task, so it does't need to add any extra time to a build finishing. The fact that edits are pushed here _before_ they're linearized in the main repo history (since they PR needs to reference a SHA) makes me even more worried that it could drift from the canonical manifest. 
> 
> And as far as b), we get that with reachability from master + gh branch protection, but I might want to go ahead and tag `vendored` for every released cockroach SHA as well, just to be safe.
</details>
Ok, sgtm

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tamird commented:
It does not use `go get`, but it's still possible for this to happen. The contents of the builders' GOPATH is really anyone's guess at this point (it is the superset of all packages ever built on them, I think). Anyway, not blocking.

---

Reviewed 8 of 9 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[CONTRIBUTING.md, line 43 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KU8oTCAxe6ScdIpCyDm:-KU8oTCAxe6ScdIpCyDn:btk04xu) ([raw file](https://github.com/cockroachdb/cockroach/blob/10bf8b08d7f78375e911003133f68f120c4b23ab/CONTRIBUTING.md#L43)):*
> ```Markdown
> 
> - see `vendor/README.md` for details on adding or updating dependencies
> - `go generate ./pkg/...` to update generated files.
> ```

Nit: `go generate ./...` is just fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KU8p6Bo7kSbsndkGrMT:bsqz892)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 15 of 16 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[CONTRIBUTING.md, line 43 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KU8oTCAxe6ScdIpCyDm:-KUA9S4iL0rzezq6Y9aw:buaufv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/10bf8b08d7f78375e911003133f68f120c4b23ab/CONTRIBUTING.md#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Nit: `go generate ./...` is just fine.
</details>
I don't think we want to run generate in our deps?

---

*[build/teamcity-checkdeps.sh, line 9 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KUA9IwTlOFHLQOsm8tQ:-KUA9IwTlOFHLQOsm8tR:b-je26lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3929de8a648cccdd38934bf37034596afb081432/build/teamcity-checkdeps.sh#L9)):*
> ```Shell
> ! git -C vendor status --porcelain | read || (git -C vendor status; git -C vendor diff -a; exit 1)
> 
> missing=$(go list -f '{{ join .Deps "\n"}}' . | grep -v "^github.com/cockroachdb/cockroach" | xargs go list -f '{{if not .Standard}}{{.ImportPath}}{{end}}')
> ```

@tamird: added this to hopefully protect against missing a dependency.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[build/teamcity-checkdeps.sh, line 9 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KUA9IwTlOFHLQOsm8tQ:-KUAC3ISSDo1uqktNQUk:b-w5mx8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3929de8a648cccdd38934bf37034596afb081432/build/teamcity-checkdeps.sh#L9)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> @tamird: added this to hopefully protect against missing a dependency.
</details>
I think this needs to be recursive. Another case is that a dependency's version is forwarded, picking up a new dependency which is missed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[build/teamcity-checkdeps.sh, line 9 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KUA9IwTlOFHLQOsm8tQ:-KUCPKi3vScWjtPq6W8U:bomtlte) ([raw file](https://github.com/cockroachdb/cockroach/blob/3929de8a648cccdd38934bf37034596afb081432/build/teamcity-checkdeps.sh#L9)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think this needs to be recursive. Another case is that a dependency's version is forwarded, picking up a new dependency which is missed.
</details>
.Deps is already recursive afaik https://golang.org/cmd/go/#hdr-List_packages which seems like it should also cover the case of a new transitive-only dep too

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tamird commented:
Whoa, TIL.

On Oct 16, 2016 08:15, "David Taylor" <notifications@github.com> wrote:

> Review status: all files reviewed at latest revision, 9 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *build/teamcity-checkdeps.sh, line 9 at r5
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KUA9IwTlOFHLQOsm8tQ:-KUCPKi3vScWjtPq6W8U:bomtlte>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/3929de8a648cccdd38934bf37034596afb081432/build/teamcity-checkdeps.sh#L9>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> I think this needs to be recursive. Another case is that a dependency's
> version is forwarded, picking up a new dependency which is missed.
>
>
> .Deps is already recursive afaik https://golang.org/cmd/go/#
> hdr-List_packages which seems like it should also cover the case of a new
> transitive-only dep too
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9900>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9900#issuecomment-254043578>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPGAa7lX-YoSEkAdzUu0WQj26wg4lks5q0hVugaJpZM4KUaUQ>
> .
>

tamird commented:
Does it include test dependencies? Docs seem incomplete.

On Oct 16, 2016 08:21, "Tamir Duberstein" <tamir@cockroachlabs.com> wrote:

> Whoa, TIL.
>
> On Oct 16, 2016 08:15, "David Taylor" <notifications@github.com> wrote:
>
>> Review status: all files reviewed at latest revision, 9 unresolved
>> discussions, all commit checks successful.
>> ------------------------------
>>
>> *build/teamcity-checkdeps.sh, line 9 at r5
>> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KUA9IwTlOFHLQOsm8tQ:-KUCPKi3vScWjtPq6W8U:bomtlte>
>> (raw file
>> <https://github.com/cockroachdb/cockroach/blob/3929de8a648cccdd38934bf37034596afb081432/build/teamcity-checkdeps.sh#L9>):*
>> *Previously, tamird (Tamir Duberstein) wrote…*
>>
>> I think this needs to be recursive. Another case is that a dependency's
>> version is forwarded, picking up a new dependency which is missed.
>>
>>
>> .Deps is already recursive afaik https://golang.org/cmd/go/#hdr
>> -List_packages which seems like it should also cover the case of a new
>> transitive-only dep too
>>
>> ------------------------------
>>
>> *Comments from Reviewable
>> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9900>*
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cockroachdb/cockroach/pull/9900#issuecomment-254043578>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABdsPGAa7lX-YoSEkAdzUu0WQj26wg4lks5q0hVugaJpZM4KUaUQ>
>> .
>>
>

dt commented:
AFAIK, no, you have to recursively use `.TestImports` for that.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KUCXJQCquInLGiCS0zB:b-rtds38)*
<!-- Sent from Reviewable.io -->

tamird commented:
Then this is indeed insufficient, isn't it?

On Oct 16, 2016 08:52, "David Taylor" <notifications@github.com> wrote:

> AFAIK, no, you have to recursively use .TestImports for that.
> ------------------------------
>
> Review status: all files reviewed at latest revision, 9 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KUCXJQCquInLGiCS0zB:b-rtds38>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9900#issuecomment-254045177>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPLy7iXDyk-kDErUc6urH7wKMBvYhks5q0h3wgaJpZM4KUaUQ>
> .
>

dt commented:
Ok, it's easy enough to add `.TestImports` and `.XTestImports` for `./pkg/...` to the starting set.

---

Review status: 15 of 16 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KUCZ6zNePuz4xUuJYvI:b-4jbd6i)*
<!-- Sent from Reviewable.io -->

dt commented:
@tamird we're not going to be go-getting things into those GOPATHs so the risk of a spuriously green build seems relatively small, plus we now checking we have all the transitive deps of any binary we're going to publish as well as our testing imports.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but i am scared

---

Reviewed 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[CONTRIBUTING.md, line 43 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KUItRv_v9yj53-lJTfS:-KUItRv_v9yj53-lJTfT:bnowr61) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb0049438292dd20382358c48649ad9e72f8841c/CONTRIBUTING.md#L43)):*
> ```Markdown
> 
> - see `vendor/README.md` for details on adding or updating dependencies
> - `go generate ./pkg/...` to update generated files.
> ```

add "run" to the beginning of this sentence for consistency

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KUIte9vr513J2ArsLIb:b-3wy1ht)*
<!-- Sent from Reviewable.io -->

tamird commented:
Retract that lgtm -- it seems that `gvt` simply punts on the problem of actually calculating dependencies. Note that a bunch of effort went into making `glock` calculate the correct set of dependencies (including those which are needed on platforms other than the current one). We need to have a solution that doesn't lose this functionality.

---

Review status: 12 of 13 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KUIyxoxJ4LoZsWqyL9P:b-cflyn6)*
<!-- Sent from Reviewable.io -->

dt commented:
What do you mean? In recursive fetch,`gvt` fetches the dep you asked it to and then parses all files, *regardless of their build tags*, for their imports and adds those to the set of required deps it needs to vendor.

---

Review status: 12 of 13 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[CONTRIBUTING.md, line 43 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KUItRv_v9yj53-lJTfS:-KUIz39Umc3iEErz6-_1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb0049438292dd20382358c48649ad9e72f8841c/CONTRIBUTING.md#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> add "run" to the beginning of this sentence for consistency
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KUJ0svOMQbSVrbJ9LsS:b-arom1y)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 12 of 13 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[build/vendor-repo.md, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDhVAU8nk7lP-UgV1:-KUJ1RgZ7jTNjbyySYsv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/vendor-repo.md#L1)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> The transition script copies this into `vendor/README` so it serves as the docs on https://github.com/cockroachdb/vendored, so I'm inclined to keep the warning until enable branch protection for `vendored` repo -- I was thinking I'd wait until we're confident this is merging to do that, but could do it now.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tamird commented:
Ah, I didn't know that. It seems that we need to update the instructions
over in https://github.com/cockroachdb/vendored to suggest "recursive fetch
of cockroach/pkg/..." instead of `gvt fetch <single-dependency>`.

On Mon, Oct 17, 2016 at 3:06 PM, David Taylor <notifications@github.com>
wrote:

> Review status: 12 of 13 files reviewed at latest revision, 9 unresolved
> discussions, some commit checks pending.
> ------------------------------
>
> *build/vendor-repo.md, line 1 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KTuDhVAU8nk7lP-UgV1:-KUJ1RgZ7jTNjbyySYsv:b-896fix>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/7adf0001291c1d0d2e448d07401203e03d30c71e/build/vendor-repo.md#L1>):*
> *Previously, dt (David Taylor) wrote…*
>
> The transition script copies this into vendor/README so it serves as the
> docs on https://github.com/cockroachdb/vendored, so I'm inclined to keep
> the warning until enable branch protection for vendored repo -- I was
> thinking I'd wait until we're confident this is merging to do that, but
> could do it now.
>
>
> Done.
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9900>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9900#issuecomment-254302471>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPErCyHn85in6eVxuJ2tfslNHQSr1ks5q08dGgaJpZM4KUaUQ>
> .
>

dt commented:
@tamird huh? I'm not sure I follow. if you want to add a new dependency `foo`, you'd `gvt fetch foo`, which would recursively add it, and everything it references (regardless of build tag) to the manifest.

tamird commented:
Why would you ever type that invocation rather than always typing `gvt
fetch github.com/cockroachdb/cockroach/pkg...`?

On Mon, Oct 17, 2016 at 3:13 PM, David Taylor <notifications@github.com>
wrote:

> @tamird <https://github.com/tamird> huh? I'm not sure I follow. if you
> want to add a new dependency foo, you'd gvt fetch foo, which would
> recursively add it, and everything it references (regardless of build tag)
> to the manifest.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9900#issuecomment-254304176>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPF1r7aJA0ZjmzGhtbS6HdZSTUYYTks5q08jEgaJpZM4KUaUQ>
> .
>

dt commented:
@tamird `gvt fetch github.com/cockroachdb/cockroach/pkg...` would, as documented, _vendor_ cockroach, which is almost certainly not what you want. Furthermore, as it is a *new* dep, you would need to fetch it before you'd have any imports of it under `pkg/...`.

tamird commented:
As discussed in person, it doesn't seem that `gvt update` does the right
thing with respect to new transitive dependencies. In fact, there is
https://github.com/FiloSottile/gvt/issues/38 that seems to be exactly this
issue.

Before this PR can merge, we need to have a reasonable workflow for dealing
with this problem.

On Mon, Oct 17, 2016 at 3:17 PM, David Taylor <notifications@github.com>
wrote:

> @tamird <https://github.com/tamird> gvt fetch github.com/cockroachdb/
> cockroach/pkg... would, as documented, *vendor* cockroach, which is
> almost certainly not what you want. Furthermore, as it is a *new* dep,
> you would need to fetch it before you'd have any imports of it under
> pkg/....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9900#issuecomment-254305220>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPJ27gZmQq2OwQEiRBjWFOnVyAo1aks5q08m9gaJpZM4KUaUQ>
> .
>

dt commented:
resolved misunderstanding of concerns with @tamird offline: while `fetch` does appears to find and manage all transitive deps, `update` doesn't pick up new transitive deps since it just bumps revs of deps in place, tracked in FiloSottile/gvt/issues/38. blocked on that, or a shell-script that recreates `update` with `fetch` calls.

dt commented:
I opened https://github.com/FiloSottile/gvt/pull/76 upstream and checked if it worked on https://github.com/dt/gvtdemo.

For now I switched the vendor README to point to that fork.

dt commented:
Rebased over #10031, and the vendor/readme suggests dt/gvt for now which has a fix for the concern @tamird raised above w.r.t. `update` handling new transitive deps. RFAL.

tamird commented:
Let's wait until we get some kind of response from upstream.

---

Reviewed 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[Makefile, line 208 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KUMKbTut_wi1q42fa2X:-KUMKbTvYfPvg1qPLbUz:b-vdf7f9) ([raw file](https://github.com/cockroachdb/cockroach/blob/27d5610737938e9783aed92c78fa21030cb70e4a/Makefile#L208)):*
> ```Makefile
> 
> GVT := ../../../../bin/gvt
> #        ^  ^  ^  ^~ GOPATH
> ```

misasligned =/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KUMyeYI77k4QuDuELMl:b-nrkfxc)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 12 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[Makefile, line 208 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KUMKbTut_wi1q42fa2X:-KUN3JyryZrvxtg7aUcy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27d5610737938e9783aed92c78fa21030cb70e4a/Makefile#L208)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> misasligned =/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Are there any licensing concerns? We would be technically redistributing the source code of these projects, I wonder if there are any extra restrictions in whatever license they may be using.

bdarnell commented:
I think any license that would place extra restrictions on source redistributions wouldn't qualify as "open source" and we'd have trouble depending on it anyway. Our current dependencies are all apache, BSD, or MIT licensed, which should definitely be safe in this respect.

dt commented:
Rebased.

After looking at gvt a little more closely, I realized it lacked the ability to remove/flatten `vendor` directories from upstream deps (looking at you, docker) so I switched this change to use glide but it doesn't change much.

tamird commented:
Can we wait and see what happens with docker? Does glide allow for **not** flattening upstream vendor dirs?

dt commented:
yes, glide allows you to specify whether or not to strip nested `vendor` dirs via the `-v` flag (it defaults to not flattening, but I flattened. because docker.)

dt commented:
I don't see a big benefit to waiting -- we should be vendoring our deps _somehow_, both to respect users' GOPATH and for reproducible builds. If/when docker figures their shit out, we can always change our handling of nested `vendor` dirs.

dt commented:
@tamird did you have any further suggested changes on this one?

tamird commented:
I don't think I've actually reviewed the latest glide-based revision. I
probably won't get to it until tomorrow.

On Mon, Nov 21, 2016 at 2:03 PM, David Taylor <notifications@github.com>
wrote:

> @tamird <https://github.com/tamird> did you have any further suggested
> changes on this one?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9900#issuecomment-262034251>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPDf7Rxssi4XNRUhaIEx2zJRNbOK0ks5rAesKgaJpZM4KUaUQ>
> .
>

tamird commented:
looks ok to me. i'm sure there are rough edges but probably time to bit the bullet.

---

Reviewed 8 of 10 files at r9.
Review status: 11 of 13 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[build/teamcity-checkdeps.sh, line 7 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KXG11p_eLMDPHXw9VBv:-KXG11p_eLMDPHXw9VBw:bskxchl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf81056311de598bad2aaf97899324b3902f789/build/teamcity-checkdeps.sh#L7)):*
> ```Shell
> 
> echo "checking that 'vendor' matches manifest"
> rm -rf vendor/*/*
> ```

how is this different from `rm -rf vendor`?

---

*[build/teamcity-checkdeps.sh, line 8 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KXG1_HrO9KLYmc7qI1S:-KXG1_HrO9KLYmc7qI1T:biv0m1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf81056311de598bad2aaf97899324b3902f789/build/teamcity-checkdeps.sh#L8)):*
> ```Shell
> echo "checking that 'vendor' matches manifest"
> rm -rf vendor/*/*
> glide install -u -v -s 2>&1
> ```

is the output merging needed, here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-:-KXG1gPx57pjrJjs3D_z:b-5k63qn)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 7 of 13 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[build/teamcity-checkdeps.sh, line 7 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KXG11p_eLMDPHXw9VBv:-KXGuL3rwCiypuMG95Rj:bmlgmtg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf81056311de598bad2aaf97899324b3902f789/build/teamcity-checkdeps.sh#L7)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how is this different from `rm -rf vendor`?
</details>
it leaves the files in the root of vendor/ alone (e.g. the README, and, importantly .git), but still clears out the mechanical stuff.

---

*[build/teamcity-checkdeps.sh, line 8 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KXG1_HrO9KLYmc7qI1S:-KXGuW9x8IxVSvxYF5jj:b4jrjwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf81056311de598bad2aaf97899324b3902f789/build/teamcity-checkdeps.sh#L8)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is the output merging needed, here?
</details>
probably not, removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r10.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[build/teamcity-checkdeps.sh, line 8 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KXG1_HrO9KLYmc7qI1S:-KXGzECQDcFRme2qVCjw:bjm8vf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf81056311de598bad2aaf97899324b3902f789/build/teamcity-checkdeps.sh#L8)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> probably not, removed.
</details>
can you use the long-form flags here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 10 of 13 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[build/teamcity-checkdeps.sh, line 8 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900#-KXG1_HrO9KLYmc7qI1S:-KXH4o4064cMiIW24Xdq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf81056311de598bad2aaf97899324b3902f789/build/teamcity-checkdeps.sh#L8)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you use the long-form flags here?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r11.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9900)*
<!-- Sent from Reviewable.io -->
